File: backend/llm/finding_prompt.py
"""User-message enrichment when the auditor clicks a known hallazgo."""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class FindingContext:
    idinventariomesdetalle: int
    idproducto: int
    idalmacen: int
    producto_nombre: str
    almacen_nombre: str
    severity_label: str | None = None


def build_finding_user_message(
    *,
    idempresa: int,
    periodo: str,
    finding: FindingContext,
    natural_question: str | None = None,
) -> str:
    """Append structured IDs so the agent calls tools without asking the auditor."""
    question = natural_question or (
        f'Explícame el hallazgo de "{finding.producto_nombre}" en el almacén '
        f'"{finding.almacen_nombre}". ¿Por qué es anómalo, qué impacto tiene '
        f"y qué debería revisar?"
    )
    sev = finding.severity_label or "—"
    return (
        f"{question}\n\n"
        "[Contexto del hallazgo — Cierre activo en pantalla]\n"
        f"- Empresa idempresa: {idempresa}\n"
        f"- Periodo: {periodo}\n"
        f"- idinventariomesdetalle: {finding.idinventariomesdetalle}\n"
        f"- idproducto: {finding.idproducto}\n"
        f"- idalmacen: {finding.idalmacen}\n"
        f"- Producto: {finding.producto_nombre}\n"
        f"- Almacén: {finding.almacen_nombre}\n"
        f"- Severidad en el ranking: {sev}\n\n"
        "Instrucción para el copiloto: el auditor ya cargó este Cierre. "
        f"Llama de inmediato get_product_history(idproducto={finding.idproducto}, "
        f"idalmacen={finding.idalmacen}). "
        "No pidas confirmación de empresa, periodo ni IDs. "
        "No ofrezcas get_top_anomalies como primer paso — este hallazgo ya está identificado."
    )


__all__ = ["FindingContext", "build_finding_user_message"]

File: frontend/src/lib/findingPrompt.ts
import type { AccionItem, Hallazgo } from "../types/cierre";

export interface FindingContextPayload {
  idinventariomesdetalle: number;
  idproducto: number;
  idalmacen: number;
  producto_nombre: string;
  almacen_nombre: string;
  severity_label?: string;
}

/** Short text shown in the chat bubble — full IDs go in finding_context. */
export function displayPromptForFinding(
  producto: string,
  almacen: string,
): string {
  return (
    `Explícame el hallazgo de "${producto}" en el almacén "${almacen}". ` +
    "¿Por qué es anómalo, qué impacto tiene y qué debería revisar?"
  );
}

export function contextFromHallazgo(a: Hallazgo): FindingContextPayload {
  return {
    idinventariomesdetalle: a.idinventariomesdetalle,
    idproducto: a.idproducto,
    idalmacen: a.idalmacen,
    producto_nombre: a.producto_nombre,
    almacen_nombre: a.almacen_nombre,
    severity_label: a.severity_label,
  };
}

/** @deprecated use contextFromHallazgo */
export const contextFromAnomaly = contextFromHallazgo;

export function contextFromBriefAction(a: AccionItem): FindingContextPayload {
  return {
    idinventariomesdetalle: a.idinventariomesdetalle,
    idproducto: a.idproducto,
    idalmacen: a.idalmacen,
    producto_nombre: a.producto_nombre,
    almacen_nombre: a.almacen_nombre,
    severity_label: a.severity_label,
  };
}
